grpclb_picker.go

```diff
@@ -0,0 +1,160 @@
+		NumCallsFinishedKnownReceived:            atomic.SwapInt64(&s.NumCallsFinishedKnownReceived, 0),
+	}
+	return stats
+}
+
+func (s *rpcStats) dropForRateLimiting() {
+	atomic.AddInt64(&s.NumCallsStarted, 1)
+	atomic.AddInt64(&s.NumCallsFinishedWithDropForRateLimiting, 1)
+	atomic.AddInt64(&s.NumCallsFinished, 1)
+}
+
+func (s *rpcStats) dropForLoadBalancing() {
+	atomic.AddInt64(&s.NumCallsStarted, 1)
+	atomic.AddInt64(&s.NumCallsFinishedWithDropForLoadBalancing, 1)
+	atomic.AddInt64(&s.NumCallsFinished, 1)
+}
+
+func (s *rpcStats) failedToSend() {
+	atomic.AddInt64(&s.NumCallsStarted, 1)
+	atomic.AddInt64(&s.NumCallsFinishedWithClientFailedToSend, 1)
+	atomic.AddInt64(&s.NumCallsFinished, 1)
+}
+
+func (s *rpcStats) knownReceived() {
+	atomic.AddInt64(&s.NumCallsStarted, 1)
+	atomic.AddInt64(&s.NumCallsFinishedKnownReceived, 1)
+	atomic.AddInt64(&s.NumCallsFinished, 1)
+}
+
+type lbPicker struct {
+	// If err is not nil, Pick always returns this err.
+	err error
+
+	mu         sync.Mutex
+	serverList []*lbpb.Server
+	nextSL     int
+	readySCs   []balancer.SubConn
+	nextSC     int
+
+	stats *rpcStats
+}
+
+// grpclb picker does two layers of picks:
+//
+// First layer: roundrobin on all servers in serverList, including drops and backends.
+// - If it picks a drop, the RPC will fail as being dropped.
+// - If it picks a backend, do a second layer pick to pick the real backend.
+//
+// Second layer: roundrobin on all backends.
+func newPicker(err error, serverList []*lbpb.Server, readySCs []balancer.SubConn, stats *rpcStats) *lbPicker {
+	grpclog.Infof("grpclb: newPicker called with: %v, %v, %v", err, serverList, readySCs)
+	if err != nil {
+		return &lbPicker{
+			err: err,
+		}
+	}
+	return &lbPicker{
+		serverList: serverList,
+		readySCs:   readySCs,
+		stats:      stats,
+	}
+}
+
+func (p *lbPicker) Pick(ctx context.Context, opts balancer.PickOptions) (conn balancer.SubConn, done func(balancer.DoneInfo), err error) {
+	if p.err != nil {
+		return nil, nil, p.err
+	}
+
+	if len(p.serverList) <= 0 && len(p.readySCs) <= 0 {
+		return nil, nil, balancer.ErrNoSubConnAvailable
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if len(p.serverList) > 0 {
+		defer func() {
+			if err != nil {
+				return
+			}
+			done = func(info balancer.DoneInfo) {
+				if !info.BytesSent {
+					p.stats.failedToSend()
+				} else if info.BytesReceived {
+					p.stats.knownReceived()
+				}
+			}
+		}()
+
+		s := p.serverList[p.nextSL]
+		p.nextSL = (p.nextSL + 1) % len(p.serverList)
+
+		// If it's a drop, return an error and fail the RPC.
+		if s.DropForRateLimiting {
+			p.stats.dropForRateLimiting()
+			return nil, nil, status.Errorf(codes.Unavailable, "request dropped by grpclb")
+		}
+		if s.DropForLoadBalancing {
+			p.stats.dropForLoadBalancing()
+			return nil, nil, status.Errorf(codes.Unavailable, "request dropped by grpclb")
+		}
+	}
+
+	// Else, do roundrobin on readySCs.
+	if len(p.readySCs) <= 0 {
```

Comment:
This can't happen unless `p.serverList > 0`, because of the check at the top.  Move this inside the if for clarity?